pkg/controller/buildrun/buildrun_controller.go

```diff
@@ -299,10 +298,14 @@ func (r *ReconcileBuildRun) retrieveTaskRun(build *buildv1alpha1.Build, buildRun
 	return nil, nil
 }
 
-func (r *ReconcileBuildRun) updateBuildRunErrorStatus(buildRun *buildv1alpha1.BuildRun, err error) *buildv1alpha1.BuildRun {
+func (r *ReconcileBuildRun) updateBuildRunErrorStatus(buildRun *buildv1alpha1.BuildRun, err error) error {
```

Comment:
`buildRun.Status.Reason` is a string, I think you want to get an `string` on this function, not an error. Whenever you call this func(), it should be passing the error message with `err.Error()`

Reply:
the func `updateBuildRunErrorStatus` report the `err` for reconcile, not for the `buildRun.Status.Reason`.

```
err = r.updateBuildRunErrorStatus(buildRun, err)
return reconcile.Result{}, err
```

So I cannot use string here.

Reply:
I didnt mean that. I meant the following:
```
func (r *ReconcileBuildRun) updateBuildRunErrorStatus(buildRun *buildv1alpha1.BuildRun, reason string) error {
	buildRun.Status.Succeeded = corev1.ConditionFalse
        buildRun.Status.Reason = reason
        buildRun.Status.StartTime = &metav1.Now()
       updateErr := r.client.Status().Update(context.TODO(), buildRun)
       return updateErr
}
```
your current function requires this.

Reply:
? I still don't get your point.

Do you mean this line `buildRun.Status.Reason = reason` ?

I wrote as `buildRun.Status.Reason = err.Error()`, and `err.Error()` is the stringm it works fine. And I suggest we set error as a common parameter and get err.Error() inside of the func so that we don't need to parse the reason out of the function and pass to the updateBuildRunErrorStatus func.

Or did I still misunderstand your point?

Reply:
```
func (r *ReconcileBuildRun) updateBuildRunErrorStatus(buildRun *buildv1alpha1.BuildRun, err error) error {
	buildRun.Status.Succeeded = corev1.ConditionFalse
	buildRun.Status.Reason = err.Error()
	now := metav1.Now()
	buildRun.Status.StartTime = &now
	updateErr := r.client.Status().Update(context.TODO(), buildRun)
	if updateErr != nil {
		return updateErr
	}
	return err
}
```

yes, please see how are you handling the `err` variable. You get the `err` as an argument to your function and then you return it from the function when the status update does not fails. You do not need to get an error, the only reason you get an error now is because you want to extract the error message out of it to put it inside the `buildRun.Status.Reason`.

What I´m trying to say is that you better get as part of the function parameters directly the error message(therefore a string) and that you only return an error , which is the one coming from the `status.update` call.

Reply:
`You do not need to get an error, the only reason you get an error now is because you want to extract the error message out of it to put it inside the buildRun.Status.Reason`

It is not true. this `error` need to be used in two places, not the only reason you said:
- one reason: set it to `buildRun.Status.Reason`
- another reason: I need to return it back to reconcile with this error if update is succeed

If I just return the error from the `status.update` call, If update status successfully, the error will be **nil**.

Then the reconcile code will become: `return reconcile.Result{}, nil`. It is not what we want.

So I have to use the error as an argument. The only concern is:
- In this functions, if we want to return only one error: validate error or update error from this function (current design)
- Or should we return two errors together back to:
```
		err = r.updateBuildRunErrorStatus(buildRun, err)
		return reconcile.Result{}, err
```
So that the recocile.Result{} report one or two errors. 

I think that is all.

You can also take a look the code which uses the `updateBuildRunErrorStatus` func. Let me know if I still understand you wrong.

Reply:
We are definitely not on the same page. Two main points:

- Getting a parameter in a function, then doing nothing with this parameter value and then returning it as part of the `return` of the function is not a desirable approach. Please change this.
- You said `another reason: I need to return it back to reconcile with this error if update is succeed` , why? I dont think you need to do this, for example:

what you do now is:
```
	err = r.client.Get(context.TODO(), types.NamespacedName{Name: buildRun.Spec.BuildRef.Name, Namespace: buildRun.Namespace}, build)
	if err != nil {
		err = r.updateBuildRunErrorStatus(buildRun, err)
		return reconcile.Result{}, err
	}
```
what you need to do is:

```
	err = r.client.Get(context.TODO(), types.NamespacedName{Name: buildRun.Spec.BuildRef.Name, Namespace: buildRun.Namespace}, build)
	if err != nil {
		errorStatus = r.updateBuildRunErrorStatus(buildRun, err)
		return reconcile.Result{}, <find a way to return 2 errors>
	}
```

this goes back to the fact that you are omitting either the error from the client get call or the error from the status update client call.  If you implement the approach where you always return two errors, then in your `updateBuildRunErrorStatus` func, you should not be passing an error, but as I said before an `string`

Reply:
After we agree to report both two validate error and update error for reconciliation, I update this `updateBuildRunErrorStatus` func that just report the updateErr, then add two errors in reconcile code, like this:
```
	if err != nil {
		updateErr := r.updateBuildRunErrorStatus(buildRun, err.Error())
		return reconcile.Result{}, fmt.Errorf("errors: %v %v", err, updateErr)
	}
```

Reply:
yes, I will open an issue for the error handling improvement, thanks